File: ordina_x/att/grader.c
// problem: ordina_x, Romeo Rizzi Mar 2015

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

extern void ordina(long int n);

#define MAX_N_BALLS 100000

static FILE *file;
static long int nBalls;
static long int order[MAX_N_BALLS];
static long int nPesate = 0;
static long int maxPesate;
static int subtask;
static int seed;


inline void swap(long int *a, long int *b) { long int tmp = *a; *a = *b; *b = tmp; }

static void generaPerm_random_uniform(long int *perm, long int n) {
   long int i;
   for(i = 0; i < n; i++)  perm[i] = i;
   for(i=n-1; i > 0; i--)
     swap( &perm[i], &perm[ (rand() % i) ] );
}


int pesa(long int bigliaA, long int bigliaB) {
  nPesate++;
  if(order[bigliaA] > order[bigliaB]) return 1;
  if(order[bigliaA] < order[bigliaB]) return -1;
  return 0;
}

void consegnaBiglieInOrdine( long int biglia_in_pos[] ) {
   long int i;
   int well_ordered = 1;
   for(i = 0; i < nBalls; i++)
     if( order[biglia_in_pos[i]] != i ) well_ordered = 0;
   fprintf(file, "%d %ld %ld\n", well_ordered, nPesate, maxPesate);
   /* only for debugging
   for(i = 0; i < nBalls; i++)
     fprintf(file, "%ld ", order[i] );
   fprintf(file, "\n");
   for(i = 0; i < nBalls; i++)
     fprintf(file, "%ld ", biglia_in_pos[i] ); */
   exit(0);
}

int main() {

#ifdef EVAL
  file = fopen("input.txt", "r");
#else
  file = stdin;
#endif

  fscanf(file, "%ld %d %d", &nBalls, &subtask, &seed);
  fclose(file);

  int LOG_UP = 1;
  long int guy = 2;
  while( guy < nBalls ) { LOG_UP++; guy *= 2; }

  srand(seed);
  generaPerm_random_uniform(order, nBalls); //genera permutazione

#ifdef EVAL
  file = fopen("output.txt", "w");
#else 
  file = stdout;
#endif

  maxPesate = nBalls*(nBalls-1)/2;
     long int i;
  if(subtask == 0)
     for(i = 0; i < nBalls; i++)  order[i] = i;
  if(subtask == 1)
     for(i = 0; i < nBalls; i++)  order[i] = nBalls-i-1;
  if(subtask == 3) { maxPesate = 5*nBalls*LOG_UP; }
  if(subtask == 4) { maxPesate = nBalls*LOG_UP; }
  if(subtask == 5) { maxPesate = nBalls*(LOG_UP -1); } 

  ordina(nBalls);

  fclose(file);
}

File: ordina_x/sol/grader.py
# problem: ordina_x, Romeo Rizzi Mar 2015

import sys

nBalls = 0
nPesate = 0
maxPesate = 0
subtask = 0
rseed = 0
outfile = None

order = None

RAND_MAX = 0x7fffffff


def rand_cp():
    global rseed
    rseed = (rseed * 1103515245 + 12345) & RAND_MAX
    return rseed


def generaPerm_random_uniform(perm, n):
    for i in range(n):
        perm[i] = i
    for i in reversed(range(1, n)):
        j = rand_cp() % i
        perm[i], perm[j] = perm[j], perm[i]


def pesa(bigliaA, bigliaB):
    global nPesate

    nPesate += 1
    if order[bigliaA] > order[bigliaB]:
        return 1
    if order[bigliaA] < order[bigliaB]:
        return -1
    return 0

def consegnaBiglieInOrdine(biglia_in_pos):
    global nBalls, order, outfile, nPesate, maxPesate
    well_ordered = True
    for i in range(nBalls):
        if order[biglia_in_pos[i]] != i:
            well_ordered = False

    print("%d %d %d" % (well_ordered, nPesate, maxPesate), file=outfile)
    # only for debugging
    # for i in range(nBalls):
    #     print(file, "%ld ", order[i])
    # print(file, "\n")
    # for i in range(nBalls):
    #   print(file, "%ld ", biglia_in_pos[i])
    sys.exit(0)


def ottieni_num_balls():
    global nBalls, rseed, outfile, maxPesate, order

    infile = open("input.txt", "r")
    # infile = sys,stdin

    (nBalls, subtask, seed) = [int(x.strip()) for x in infile.read().split()]
    infile.close()

    LOG_UP = 1
    guy = 2
    while guy < nBalls:
        LOG_UP += 1
        guy *= 2

    order = [0] * nBalls

    rseed = seed
    generaPerm_random_uniform(order, nBalls)  # genera permutazione

    outfile = open("output.txt", "w")
    # outfile = sys.stdout

    maxPesate = nBalls * (nBalls - 1) // 2

    if subtask == 0:
        for i in range(nBalls):
            order[i] = i
    if subtask == 1:
        for i in range(nBalls):
            order[i] = nBalls - i - 1
    if subtask == 3:
        maxPesate = 5 * nBalls * LOG_UP
    if subtask == 4:
        maxPesate = nBalls * LOG_UP
    if subtask == 5:
        maxPesate = nBalls * (LOG_UP - 1)

    return nBalls